smtlib-backends-z3/src/SMTLIB/Backends/Z3.hs

```diff
@@ -31,32 +33,53 @@ import qualified Language.C.Inline.Unsafe as CU
 import qualified Language.C.Types as C
 import SMTLIB.Backends (Backend (..))
 
-data LogicalContext
+data Z3Context
+
+data Z3Config
 
 C.context
   ( C.baseCtx
       <> C.fptrCtx
       <> C.bsCtx
       <> mempty
         { C.ctxTypesTable =
-            M.singleton (C.TypeName "Z3_context") [t|Ptr LogicalContext|]
+            M.singleton (C.TypeName "Z3_context") [t|Ptr Z3Context|]
+              <> M.singleton (C.TypeName "Z3_config") [t|Ptr Z3Config|]
         }
   )
 C.include "z3.h"
 
+data Config = Config
+  { -- | A list of options to set during the solver's initialization.
+    -- Each pair is of the form @(paramId, paramValue)@, e.g.
+    -- @(":produce-models", "true")@.
+    parameters :: [(BS.ByteString, BS.ByteString)]
+  }
+
+-- | By default, don't set any options during initialization.
+instance Default Config where
+  def = Config []
+
 data Handle = Handle
   { -- | A black-box representing the internal state of the solver.
-    context :: ForeignPtr LogicalContext
+    context :: ForeignPtr Z3Context
   }
 
 -- | Create a new solver instance.
-new :: IO Handle
-new = do
+new :: Config -> IO Handle
+new config = do
   let ctxFinalizer =
         [C.funPtr| void free_context(Z3_context ctx) {
                  Z3_del_context(ctx);
                  } |]
-
+  -- we don't set a finalizer for this object as we manually delete it during the
+  -- context's creation
+  cfg <- [CU.exp| Z3_config { Z3_mk_config() } |]
+  flip mapM_ (parameters config) $ \(paramId, paramValue) ->
+    [CU.exp| void { Z3_set_param_value($(Z3_config cfg),
+                                       $bs-ptr:paramId,
+                                       $bs-ptr:paramValue)
+           } |]
```

Comment:
One thing I'm wondering is if there's a better way to do this, as it forces the introduction of the `Z3Config` datatype that we wouldn't need otherwise.

Reply:
You could `unzip (parameters config)` convert the resulting lists to C arrays, and create the config and the context in a single `inline-c` block. But it is fine having a `Z3Config` type in Haskell too.